### ISSUE TITLE
Migrations performed by Flyway or Liquibase may not have completed before the database is accessed via jOOQ

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#25055]
If jOOQ is used to interact with a database during application context refresh, migrations performed by Flyway or Liquibase may not have completed. This can happen because there's no guarantee that the Flyway or Liquibase beans will have been created and initialised before any jOOQ beans are created and used. I think we need some `AbstractDependsOnBeanFactoryPostProcessor` beans so that any jOOQ `DSLContext` beans depend on the Flyway and Liquibase beans.

**Comment by lukaseder** [comment]
Interesting, could this be a more general problem? Would a migration-aware `DataSource` (or connection pool) be interesting, with a mutex that prevents migrations and ordinary access from happening at the same time?

**Comment by wilkinsona** [comment]
It is a general problem and we have a (fairly) general solution for it. There's a bug here as we've neglected to use that solution for jOOQ's `DSLContext`. I noticed this as part of a broader reworking of DataSource initialization for Boot 2.5. Part of that may result in a further generalisation of the current solution so that it's more extensible.

>  a mutex that prevents migrations and ordinary access from happening at the same time

Other than one scenario involved deferred bootstrapping of JPA, everything is single threaded. We need to ensure that things happen in the right order rather than preventing them from happening in parallel.

**Comment by lukaseder** [comment]
I see, thanks for the explanation!

**Comment by wilkinsona** [comment]
Closing in favour of #25279.

**Comment by neiser** [comment]
@wilkinsona I've found this issue while using Spring Boot 2.7.3 probably suffering exactly from the problem that Flyway might not have run during application startup but already using jOOQ's `DSLContext` within a `@Scheduled` method. It then fails with an error telling me the database table isn't present. I've seen this fix has been [forward-ported to 2.4.x](https://github.com/spring-projects/spring-boot/blob/2.4.x/spring-boot-project/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/flyway/FlywayAutoConfiguration.java#L37), but I can't tell how it's now solved from 2.5.x onwards. Searching `org.springframework.boot.autoconfigure.jooq` package for keyword `flyway` and `org.springframework.boot.autoconfigure.flyway` for keyword `jooq` makes me wonder how it's ensured that jOOQ's `DSLContext` bean is only available once Flyway's migration has happened. Can you enlighten me? Or is this a bug which has now re-appeared after the refactoring of DataSources in 2.5.x you've mentioned above?

**Comment by wilkinsona** [comment]
`org.springframework.boot.jooq.JooqDependsOnDatabaseInitializationDetector` should ensure that any `DSLContext` beans depend upon any beans that are responsible for `DataSource` initialization. `Flyway` is one such type of bean as identified by `org.springframework.boot.flyway.FlywayDatabaseInitializerDetector`.

**Comment by neiser** [comment]
@wilkinsona Thanks a lot for this immediate answer. I'll see if I can reproduce it now that I believe jOOQ is indeed later available than Flyway :+1:
